src/smartpeak/source/core/ProgessInfo.cpp

```diff
@@ -0,0 +1,187 @@
+                                        running_commands_.end(), 
+                                        std::make_tuple(command_index, command_name)), 
+                                        running_commands_.end());
+    commands_step_++;
+  }
+
+  void ProgressInfo::onApplicationProcessorEnd()
+  {
+    last_run_time_ = runningTime();
+    running_batch_ = nullptr;
+    running_ = false;
+  }
+
+  void ProgressInfo::onSequenceProcessorStart(const size_t nb_injections)
+  {
+    running_batch_ = std::make_shared<RunningBatch>(nb_injections);
+  }
+
+  void ProgressInfo::onSequenceProcessorSampleStart(const std::string& sample_name)
+  {
+    running_batch_->running_items_.push_back(sample_name);
+  }
+
+  void ProgressInfo::onSequenceProcessorSampleEnd(const std::string& sample_name)
+  {
+    onRunningBatchEndItem(sample_name);
+  }
+
+  void ProgressInfo::onSequenceProcessorEnd()
+  {
+  }
+
+  void ProgressInfo::onSequenceSegmentProcessorStart(const size_t nb_segments)
+  {
+    running_batch_ = std::make_shared<RunningBatch>(nb_segments);
+  }
+
+  void ProgressInfo::onSequenceSegmentProcessorSampleStart(const std::string& segment_name)
+  {
+    running_batch_->running_items_.push_back(segment_name);
+  }
+
+  void ProgressInfo::onSequenceSegmentProcessorSampleEnd(const std::string& segment_name)
+  {
+    onRunningBatchEndItem(segment_name);
+  }
+
+  void ProgressInfo::onSequenceSegmentProcessorEnd()
+  {
+  }
+
+  void ProgressInfo::onSampleGroupProcessorStart(const size_t nb_groups)
+  {
+    running_batch_ = std::make_shared<RunningBatch>(nb_groups);
+  }
+
+  void ProgressInfo::onSampleGroupProcessorSampleStart(const std::string& group_name)
+  {
+    running_batch_->running_items_.push_back(group_name);
+  }
+
+  void ProgressInfo::onSampleGroupProcessorSampleEnd(const std::string& group_name)
+  {
+    onRunningBatchEndItem(group_name);
+  }
+
+  void ProgressInfo::onSampleGroupProcessorEnd()
+  {
+  }
+
+  void ProgressInfo::onRunningBatchEndItem(const std::string& item_name)
+  {
+    running_batch_->running_items_.erase(std::remove(running_batch_->running_items_.begin(),
+      running_batch_->running_items_.end(),
+      item_name),
+      running_batch_->running_items_.end());
+    running_batch_->current_step_++;
+    running_batch_->last_item_end_time_point_ = std::chrono::steady_clock::now();
+  }
+
+  std::chrono::steady_clock::duration ProgressInfo::runningTime() const
+  {
+    if (isRunning())
+    {
+      return std::chrono::steady_clock::now() - application_processor_start_time_;
+    }
+    else
+    {
+      return std::chrono::steady_clock::duration::zero();
+    }
+  }
+
+  std::optional<std::chrono::steady_clock::duration> ProgressInfo::estimatedRemainingTime() const
+  {
+    if (running_batch_  && running_batch_->current_step_ > 0)
+    {
+      auto time_per_steps = ((running_batch_->last_item_end_time_point_ - application_processor_start_time_) / running_batch_->current_step_);
```

Comment:
yes, we can do better :)

Reply:
time_factor (some kind of wieght) could be a solution indeed. we may need some metrics as well. there are ways let see how it would be useful for the users!

Reply:
Very true.  Let's not spend too much time on this as we have not validated what is most useful to the users at this stage yet.